language/borrow-graph/src/references/mod.rs

```diff
@@ -0,0 +1,160 @@
+
+impl RefID {
+    pub const fn new(x: usize) -> Self {
+        RefID(x)
+    }
+}
+
+#[derive(Clone)]
+pub(crate) struct BorrowEdge<Loc: Copy, Lbl: Clone + Eq> {
+    pub(crate) strong: bool,
+    pub(crate) path: Path<Lbl>,
+    pub(crate) loc: Loc,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub(crate) struct BorrowEdges<Loc: Copy, Lbl: Clone + Eq>(
+    pub(crate) BTreeMap<RefID, BTreeSet<BorrowEdge<Loc, Lbl>>>,
+);
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub(crate) struct Ref<Loc: Copy, Lbl: Clone + Eq> {
+    // x is borrowed by y
+    pub(crate) borrowed_by: BorrowEdges<Loc, Lbl>,
+    // y borrows from x
+    pub(crate) borrows_from: BTreeSet<RefID>,
+    // true if mutable, false otherwise
+    pub(crate) mutable: bool,
+}
+
+//**************************************************************************************************
+// Impls
+//**************************************************************************************************
+
+impl<Loc: Copy, Lbl: Clone + Eq> BorrowEdges<Loc, Lbl> {
+    pub(crate) fn new() -> Self {
+        Self(BTreeMap::new())
+    }
+}
+
+impl<Loc: Copy, Lbl: Clone + Eq> Ref<Loc, Lbl> {
+    pub(crate) fn new(mutable: bool) -> Self {
+        let borrowed_by = BorrowEdges::new();
+        let borrows_from = BTreeSet::new();
+        Self {
+            borrowed_by,
+            borrows_from,
+            mutable,
+        }
+    }
+}
+
+//**********************************************************************************************
+// Remap
+//**********************************************************************************************
+
+impl<Loc: Copy, Lbl: Clone + Eq> BorrowEdges<Loc, Lbl> {
+    pub(crate) fn remap_refs(&mut self, id_map: &BTreeMap<RefID, RefID>) {
+        for (old, new) in id_map {
+            if let Some(edges) = self.0.remove(old) {
+                self.0.insert(*new, edges);
+            }
+        }
+    }
+}
+
+impl<Loc: Copy, Lbl: Clone + Eq> Ref<Loc, Lbl> {
+    pub(crate) fn remap_refs(&mut self, id_map: &BTreeMap<RefID, RefID>) {
+        self.borrowed_by.remap_refs(id_map);
+        remap_set(&mut self.borrows_from, id_map)
+    }
+}
+
+//**********************************************************************************************
+// Joins
+//**********************************************************************************************
+
+impl<Loc: Copy, Lbl: Clone + Eq + Ord> BorrowEdges<Loc, Lbl> {
+    pub(crate) fn join(&self, other: &Self) -> Self {
+        Self(union_map(&self.0, &other.0, |_, s1, s2| union_sets(s1, s2)))
+    }
+}
+
+impl<Loc: Copy, Lbl: Clone + Eq + Ord> Ref<Loc, Lbl> {
+    pub(crate) fn join(&self, other: &Self) -> Self {
+        let borrowed_by = self.borrowed_by.join(&other.borrowed_by);
+        let borrows_from = union_sets(&self.borrows_from, &other.borrows_from);
+        assert!(
+            self.mutable == other.mutable,
+            "ICE type checking or freezing failed"
+        );
+        let mutable = self.mutable;
```

Comment:
Spurious/stupid question: Why isn't a non-mutable refrence promoted to mutable? (i.e. we have join(mutable, immutable) = mutable)?

Reply:
At this point in the checks, I've removed subtyping and added explicit freezing if needed